src/System.Linq.Expressions/src/System/Linq/Expressions/StackGuard.cs

```diff
@@ -0,0 +1,91 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Linq.Expressions
+{
+    internal sealed class StackGuard
+    {
+        private const int StackProbingThreshold = 100;
+        private const int MaxExecutionStackCount = 1024;
+
+        private int _recursionDepth;
+        private int _executionStackCount;
+
+        public bool TryEnterOnCurrentStack()
+        {
+            if (++_recursionDepth >= StackProbingThreshold)
+            {
+                try
+                {
+                    RuntimeHelpers.EnsureSufficientExecutionStack();
+                }
+                catch (InsufficientExecutionStackException) when (_executionStackCount < MaxExecutionStackCount)
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        public void Exit()
+        {
+            _recursionDepth--;
+        }
+
+        public void RunOnEmptyStack(Action<object> action, object state)
+        {
```

Comment:
Yeah, to my knowledge there's not a great way to do that.  But, other than for usability, does it really matter?  This should be a super rare case where if we hit it we're already about to blow the stack and have and will continue to pay lots of costs, right?  In such a case, hopefully an extra object allocation won't matter.  Besides, the call sites are all already allocating a tuple, so using a TState here wouldn't actually save any allocations (I agree it would help a bit with usability, as the call site could then avoid a cast).